File: Eino testing/pdf_processor.py
"""
PDF processing using pdfplumber for parsing energy advice documents.
"""

import os
from pathlib import Path
from typing import List, Dict, Any
import tiktoken

try:
    import pdfplumber
    PDFPLUMBER_AVAILABLE = True
except ImportError:
    PDFPLUMBER_AVAILABLE = False
    print("[WARN] pdfplumber ikke installert. Bruk: pip install pdfplumber")


def count_tokens(text: str) -> int:
    """Count tokens in text using tiktoken."""
    try:
        encoding = tiktoken.encoding_for_model("gpt-4")
        return len(encoding.encode(text))
    except:
        # Fallback: rough estimate
        return len(text.split()) * 1.3


def chunk_text(text: str, chunk_size: int = 200, overlap: int = 50) -> List[str]:
    """
    Split text into overlapping chunks.
    
    Args:
        text: Text to chunk
        chunk_size: Target chunk size in tokens
        overlap: Overlap size in tokens
    
    Returns:
        List of text chunks
    """
    if not text.strip():
        return []
    
    # Split by paragraphs first
    paragraphs = [p.strip() for p in text.split('\n\n') if p.strip()]
    
    chunks = []
    current_chunk = []
    current_tokens = 0
    
    for para in paragraphs:
        para_tokens = count_tokens(para)
        
        # If paragraph is too large, split it
        if para_tokens > chunk_size:
            # Add current chunk if exists
            if current_chunk:
                chunks.append('\n\n'.join(current_chunk))
                current_chunk = []
                current_tokens = 0
            
            # Split large paragraph by sentences
            sentences = para.split('. ')
            for sent in sentences:
                sent_tokens = count_tokens(sent)
                if current_tokens + sent_tokens > chunk_size and current_chunk:
                    chunks.append('\n\n'.join(current_chunk))
                    # Keep overlap
                    overlap_text = '\n\n'.join(current_chunk[-overlap//20:])
                    current_chunk = [overlap_text, sent] if overlap_text else [sent]
                    current_tokens = count_tokens('\n\n'.join(current_chunk))
                else:
                    current_chunk.append(sent)
                    current_tokens += sent_tokens
        else:
            # Check if adding paragraph exceeds chunk size
            if current_tokens + para_tokens > chunk_size and current_chunk:
                chunks.append('\n\n'.join(current_chunk))
                # Keep overlap
                overlap_text = '\n\n'.join(current_chunk[-overlap//20:])
                current_chunk = [overlap_text, para] if overlap_text else [para]
                current_tokens = count_tokens('\n\n'.join(current_chunk))
            else:
                current_chunk.append(para)
                current_tokens += para_tokens
    
    # Add final chunk
    if current_chunk:
        chunks.append('\n\n'.join(current_chunk))
    
    return chunks


def extract_text_from_pdf(pdf_path: str) -> Dict[str, Any]:
    """
    Extract text from PDF using pdfplumber.
    
    Returns:
        Dictionary with 'text', 'pages', 'metadata'
    """
    if not PDFPLUMBER_AVAILABLE:
        raise ImportError("pdfplumber ikke installert. Installer med: pip install pdfplumber")
    
    pdf_path_obj = Path(pdf_path)
    if not pdf_path_obj.exists():
        raise FileNotFoundError(f"PDF ikke funnet: {pdf_path}")
    
    text_parts = []
    pages = []
    
    try:
        with pdfplumber.open(pdf_path) as pdf:
            for page_num, page in enumerate(pdf.pages, start=1):
                # Extract text from page
                page_text = page.extract_text()
                
                if page_text and page_text.strip():
                    text_parts.append(page_text)
                    pages.append(page_num)
    except Exception as e:
        raise Exception(f"Feil ved PDF-behandling med pdfplumber: {str(e)}")
    
    full_text = '\n\n'.join(text_parts)
    
    return {
        "text": full_text,
        "pages": pages,
        "metadata": {
            "source_file": pdf_path_obj.name,
            "file_path": str(pdf_path),
            "num_pages": len(pages),
            "doc_type": "energy_advice"
        }
    }


def process_pdf_directory(pdf_dir: str = "pdf", chunk_size: int = 200, 
                          chunk_overlap: int = 50) -> List[Dict[str, Any]]:
    """
    Process all PDFs in directory and return chunks.
    
    Returns:
        List of chunk dictionaries with text, metadata, and source info
    """
    pdf_dir_path = Path(pdf_dir)
    if not pdf_dir_path.exists():
        raise FileNotFoundError(f"PDF-mappe ikke funnet: {pdf_dir}")
    
    all_chunks = []
    
    # Find all PDF files
    pdf_files = list(pdf_dir_path.glob("*.pdf"))
    
    if not pdf_files:
        print(f"[WARN] Ingen PDF-filer funnet i {pdf_dir}")
        return []
    
    print(f"[INFO] Fant {len(pdf_files)} PDF-filer")
    
    for pdf_file in pdf_files:
        try:
            print(f"[INFO] Behandler {pdf_file.name}...")
            result = extract_text_from_pdf(str(pdf_file))
            
            if not result["text"].strip():
                print(f"[WARN] {pdf_file.name}: Ingen tekst funnet")
                continue
            
            # Chunk the text
            chunks = chunk_text(result["text"], chunk_size, chunk_overlap)
            
            # Create chunk entries with metadata
            for i, chunk_text_content in enumerate(chunks):
                # Estimate page number (rough)
                page_estimate = None
                if result["pages"]:
                    page_estimate = result["pages"][min(
                        i * len(result["pages"]) // len(chunks),
                        len(result["pages"]) - 1
                    )]
                
                chunk_entry = {
                    "text": chunk_text_content,
                    "metadata": {
                        **result["metadata"],
                        "chunk_index": i,
                        "total_chunks": len(chunks),
                        "page": page_estimate,
                        "tokens": count_tokens(chunk_text_content)
                    }
                }
                all_chunks.append(chunk_entry)
            
            print(f"[OK] {pdf_file.name}: {len(chunks)} chunks")
            
        except Exception as e:
            print(f"[ERROR] Feil ved behandling av {pdf_file.name}: {e}")
            continue
    
    print(f"[OK] Totalt {len(all_chunks)} chunks generert")
    return all_chunks


if __name__ == "__main__":
    # Test PDF processing
    if PDFPLUMBER_AVAILABLE:
        chunks = process_pdf_directory()
        print(f"\nFørste chunk eksempel:")
        if chunks:
            print(chunks[0]["text"][:500])
            print(f"\nMetadata: {chunks[0]['metadata']}")
    else:
        print("pdfplumber ikke tilgjengelig. Installer med: pip install pdfplumber")

File: Eino testing/backend/pdf_processor.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os/exec"
	"path/filepath"
	"strings"
)

type DocumentChunk struct {
	Text     string
	Metadata map[string]interface{}
}

func processPDFDirectory(pdfDir string) ([]DocumentChunk, error) {
	chunks := []DocumentChunk{}

	// Find all PDF files
	pdfFiles, err := filepath.Glob(filepath.Join(pdfDir, "*.pdf"))
	if err != nil {
		return nil, fmt.Errorf("failed to find PDF files: %w", err)
	}

	if len(pdfFiles) == 0 {
		log.Println("No PDF files found in", pdfDir)
		return chunks, nil
	}

	log.Printf("Found %d PDF files", len(pdfFiles))

	for _, pdfPath := range pdfFiles {
		fileName := filepath.Base(pdfPath)
		log.Printf("Processing %s...", fileName)

		// Use Python pdfplumber for PDF extraction (more reliable)
		// This calls a Python script that uses pdfplumber
		text, err := extractPDFTextWithPython(pdfPath)
		if err != nil {
			log.Printf("Warning: failed to extract text from %s: %v", fileName, err)
			continue
		}

		if strings.TrimSpace(text) == "" {
			log.Printf("Warning: no text extracted from %s", fileName)
			continue
		}

		// Chunk the text
		textChunks := chunkText(text, ChunkSize, ChunkOverlap)

		// Create chunk entries
		for i, chunkText := range textChunks {
			chunks = append(chunks, DocumentChunk{
				Text: chunkText,
				Metadata: map[string]interface{}{
					"source_file":  fileName,
					"chunk_index":  i,
					"total_chunks": len(textChunks),
					"doc_type":     "energy_advice",
				},
			})
		}

		log.Printf("Processed %s: %d chunks", fileName, len(textChunks))
	}

	log.Printf("Total: %d chunks from %d PDFs", len(chunks), len(pdfFiles))
	return chunks, nil
}

func chunkText(text string, chunkSize, overlap int) []string {
	if text == "" {
		return []string{}
	}

	// Split by paragraphs
	paragraphs := strings.Split(text, "\n\n")
	var chunks []string
	currentChunk := ""
	currentLength := 0

	for _, para := range paragraphs {
		para = strings.TrimSpace(para)
		if para == "" {
			continue
		}

		paraLength := len(strings.Fields(para)) // Approximate token count

		if currentLength+paraLength > chunkSize && currentChunk != "" {
			chunks = append(chunks, currentChunk)

			// Keep overlap
			if overlap > 0 && len(chunks) > 0 {
				words := strings.Fields(currentChunk)
				if len(words) > overlap {
					currentChunk = strings.Join(words[len(words)-overlap:], " ")
					currentLength = overlap
				} else {
					currentChunk = ""
					currentLength = 0
				}
			} else {
				currentChunk = ""
				currentLength = 0
			}
		}

		if currentChunk != "" {
			currentChunk += "\n\n"
		}
		currentChunk += para
		currentLength += paraLength
	}

	if currentChunk != "" {
		chunks = append(chunks, currentChunk)
	}

	return chunks
}

// extractPDFTextWithPython uses Python pdfplumber to extract text
func extractPDFTextWithPython(pdfPath string) (string, error) {
	// Create a temporary Python script
	pythonScript := `
import sys
import pdfplumber
import json

pdf_path = sys.argv[1]
text_parts = []

try:
    with pdfplumber.open(pdf_path) as pdf:
        for page in pdf.pages:
            page_text = page.extract_text()
            if page_text and page_text.strip():
                text_parts.append(page_text)
    print(json.dumps({"text": "\n\n".join(text_parts)}))
except Exception as e:
    print(json.dumps({"error": str(e)}))
    sys.exit(1)
`

	// Execute Python script
	cmd := exec.Command("python", "-c", pythonScript, pdfPath)
	output, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("python execution failed: %w", err)
	}

	// Parse JSON output
	var result struct {
		Text  string `json:"text"`
		Error string `json:"error"`
	}
	if err := json.Unmarshal(output, &result); err != nil {
		return "", fmt.Errorf("failed to parse python output: %w", err)
	}

	if result.Error != "" {
		return "", fmt.Errorf("python error: %s", result.Error)
	}

	return result.Text, nil
}
